### ISSUE TITLE
Add watch mode for deno check

### ISSUE THREAD

**Issue by bxantus** [denoland/deno#14858]
For me it would make a lot of sense to run `check` in watch mode. 
I think that adding check subcommand was a great idea, and running it in watch mode would provide the best development experience (at least for me). Also  if I'm not mistaken it would check further edits faster, as typescript compiler is already warmed up after the first check.

```
deno check --watch
```

**Comment by bartlomieju** [comment]
> For me it would make a lot of sense to run check in watch mode.
I think that adding check subcommand was a great idea, and running it in watch mode would provide the best development experience (at least for me).

Sure, makes sense; it should be fairly easy to implement.

> Also if I'm not mistaken it would check further edits faster, as typescript compiler is already warmed up after the first check.

Not really - to do so, we'd need to rearchitecture the `check` command for `--watch` mode - that's because, currently `--watch` literally restarts the whole subcommand that it's watching - there's no state preserved between the runs. To implement what you're suggesting we'd instead have to keep a TSC isolate running (and to do that we'd probably need to implement a new TS host or reuse one used in the `deno lsp`). @kitsonk any thoughts on this?

**Comment by ChristianSiegert** [comment]
@bxantus, does `deno run --check --watch` do what you want?

**Comment by bxantus** [comment]
No, it isn't the same. 
Currently I'm using Deno for creating web apps (running in the browser). I find it much easier to use Deno tooling compared to node.js. So `deno run` would not work in this case. I would like to use `deno check --watch` like a compiler (the typescript compiler), checking if everything is ok. Then I bundle the scripts using esbuild.

Even if I developed a console app (running it with deno run), using `deno check --watch` would suit me better. As I want to check my code frequently, but run it only when I'm fully finished with my modifications.
That's why I'm in favor of removing checking by default from run, and using checks mostly while developing.

**Comment by bartlomieju** [comment]
Marked as "good first issue", since the naive approach should be very easy to implement based on existing subcommands with support for --watch.

**Comment by VishnuJin** [comment]
@bartlomieju can you please assign this to me? I'll give it a try!

**Comment by bartlomieju** [comment]
> @bartlomieju can you please assign this to me? I'll give it a try!

I cannot assign you the issue, but feel free to pick it up.

**Comment by cuobiezi** [comment]
Can I take this one? I will try to finish is before next Monday.

**Comment by bartlomieju** [comment]
@cuobiezi please do

**Comment by cuobiezi** [comment]
@bartlomieju  please refer #15325, But I don't know  whether unit test is necessary , And I have no idea how to write unit test .I only tested it by manually LOL.

**Comment by VishnuJin** [comment]
@cuobiezi thanks for taking up the PR

I just tried to run check with watch args and have the following doubt

- after watch is run it is expected to catch and display the errors to right, but nothing changes in the watch windows

#### Start point - all good
![IMG_20220728_222758.jpg](https://user-images.githubusercontent.com/45007338/181595248-2788b626-6f76-4397-ac20-e7a61f947655.jpg)

### Expected error - but not update
![IMG_20220728_222728.jpg](https://user-images.githubusercontent.com/45007338/181595435-3a442f1f-dd72-469c-84ed-16ca4e586f16.jpg)

Am I missing something here ?

**Comment by the-nikhil-jugdan** [comment]
Is this issue still open ? I would like pick it up

**Comment by kamilogorek** [comment]
It appears to be implemented in https://github.com/denoland/deno/pull/15325 but is not merged yet.

**Comment by the-nikhil-jugdan** [comment]
Okay, thanks for the info

**Comment by niedzielski** [comment]
Now that the bundle command is deprecated, I think this feature will be sorely wanted. I can bundle with `esbuild` instead which does have a watch API but to type-check, I think I still need something with import-aware watch abilities.

**Comment by vaishnavmhetre** [comment]
Seems like PR https://github.com/denoland/deno/pull/15325, resolve this.

**Comment by jsejcksn** [comment]
> Now that the bundle command is deprecated, I think this feature will be sorely wanted. I can bundle with `esbuild` instead which does have a watch API but to type-check, I think I still need something with import-aware watch abilities.

@niedzielski I've advocated for a runtime type-checking API in [#15478](https://github.com/denoland/deno/issues/15478), which would be very helpful in the scenario that you described.

**Comment by birgersp** [comment]
> It appears to be implemented in #15325 but is not merged yet.

the PR is deleted 😟

**Comment by forest1102** [comment]
Any progress here?

**Comment by Cethrivias** [comment]
Are there plans to implement it in any capacity? Type checking without running the app is pretty useful

**Comment by birgersp** [comment]
> Are there plans to implement it in any capacity? Type checking without running the app is pretty useful

Use `deno check` to check types without running the app

https://docs.deno.com/runtime/reference/cli/check/

**Comment by Cethrivias** [comment]
> Use `deno check` to check types without running the app

Yes that is exactly what I want to do, but I want ‘—watch’ flag for it. I think this is what this whole thread is about

**Comment by birgersp** [comment]
> > Use `deno check` to check types without running the app
> 
> Yes that is exactly what I want to do, but I want `—watch` flag for it. I think this is what this whole thread is about

Sorry, my bad. I didn't understand the context. 🤦

**Comment by 0x009922** [comment]
By the way, there is a slightly older duplicate: #9724

**Comment by 0x009922** [comment]
I would like to work on this issue.

In fact, I've already done some research and experiments. I found a few things:

- Different Deno commands handle `--watch` quite differently. I mean the way they collect files to watch.
- `deno fmt --watch` and `deno lint --watch` seem to not notice appearance of new files. Will open an issue when confirmed.
- `deno run --watch` and `deno serve --watch` have redundant `watcher_communicator.watch_paths(..)`:
  https://github.com/denoland/deno/blob/ee4c14a550ff65d1982425922cf116cda0066954/cli/tools/run/mod.rs#L144
  https://github.com/denoland/deno/blob/ee4c14a550ff65d1982425922cf116cda0066954/cli/tools/serve.rs#L167
  This is effectively ignored and overwritten later by paths from `FileWatcherReporter`:
  https://github.com/denoland/deno/blob/ee4c14a550ff65d1982425922cf116cda0066954/cli/graph_util.rs#L1282-L1285
  Maybe I will open a small PR with removal of this, when confirm the tests pass.

Currently I am focusing on the question: what is the proper way to collect watch files for `deno check --watch`.

**Comment by 0x009922** [comment]
> This is the one and only reason I haven't moved my projects to deno yet. I really love ALL the features deno offers (and thank you very much for the work and effort you are putting into it) but running the compiler in watch mode (`tsc --watch`) on the side is such a powerful tool to be forced to give up.

You can still use a generic watch tool to achieve a similar result:

```sh
watchexec -e ts,js,json deno check
```

Though `deno check --watch` would be able to do thing more optimally.

**Comment by 0x009922** [comment]
I am still working on this issue, but found it to be larger than I though: #29132

I would prefer to solve that one first. Without it, it seems, the only way to implement `deno check --watch` is to produce even more ad-hoc code than it is already there.
